### DISCUSSION TITLE
Record type validation

### DISCUSSION THREAD

**Discussion by foxesknow** [dotnet/csharplang#5320] [Language Ideas]
There's no easy way to validate the parameters passed to a record type without having to manually implement the `init` property and add checking:

```
record Person
{
    private readonly int _age;
    
    public int Age
    {
        get => _age;
        init =>
        {
            if (value < 0)
            {
                throw new ArgumentException("Argument cannot be negative.", nameof(value));
            }
            _age = value;
        }
    }
}
```
It would be useful to be able to have some sort of post construction phase which allows the record to validate itself in one place (much like a constructor would), rather than scatter the validation throughout the class and lose the readability of auto properties.

**Comment by jnm2** [comment]
Yes, this is one of the things that often means I can't benefit from records.
https://github.com/jaredpar/csharplang/blob/record/proposals/recordsv3.md#validators
https://github.com/dotnet/csharplang/discussions/3323 and https://github.com/dotnet/csharplang/discussions/3407 (meeting notes discussions)
https://github.com/dotnet/csharplang/discussions/4002 (duplicate)

**Comment by RikkiGibson** [comment]
It feels like `field` (#140) may help with this.

**Reply by jnm2** [reply]
That could get it down to this (and both could be refactored to an expression form):

```cs
record Person
{    
    public int Age
    {
        get;
        init
        {
            if (value < 0)
            {
                throw new ArgumentException("Argument cannot be negative.", nameof(value));
            }
            field = value;
        }
    }
}
```

**Reply by jnm2** [reply]
Together with https://github.com/safakgur/guard:

```cs
record Person
{    
    public int Age { get; init => field = Guard.Argument(value, nameof(value)).NotNegative(); }
}
```

**Reply by TonyValenti** [reply]
I often create a method named "validate" which returns an enumerable of exception that has all validation errors.

**Comment by Joe4evr** [comment]
The language team [has already shown interest in having something like this](https://youtu.be/fIXujTOA6BE?t=2749), and not even restricted to record types.
